File: problem/Max_Product/maxProduct.c
long long getSum(struct TreeNode *root)
{
    if (!root)
        return 0;
    long long sum = root->val + getSum(root->left) + getSum(root->right);
    root->val = (int)sum;
    return sum;
}

void findMax(struct TreeNode *root, long long total, long long *ans)
{
    if (!root)
        return;

    long long sub = (long long)root->val;
    long long cur = (total - sub) * sub;
    if (cur > *ans)
        *ans = cur;

    findMax(root->left, total, ans);
    findMax(root->right, total, ans);
}

int maxProduct(struct TreeNode *root)
{
    long long ans = 0;
    long long total = getSum(root);

    findMax(root, total, &ans);

    return (int)(ans % 1000000007);
}

File: problem/Max_Product/maxProduct.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

var maxProduct = function(root) {
    const MOD = 1e9 + 7;

    // DFS to compute subtree sums (in-place)
    const dfs = (node) => {
        if (!node) return 0;
        node.val += dfs(node.left) + dfs(node.right);
        return node.val;
    };

    const total = dfs(root);

    let ans = 0;
    const queue = [root];

    // BFS to compute max product
    while (queue.length > 0) {
        const node = queue.shift();
        if (!node) continue;

        const cur = (total - node.val) * node.val;
        ans = Math.max(ans, cur);

        if (node.left) queue.push(node.left);
        if (node.right) queue.push(node.right);
    }

    return ans % MOD;
};